File: test/jte-runtime-cp-test-gradle-newtask/src/test/java/gg/jte/TemplateEngineTest.java
package gg.jte;

import gg.jte.output.StringOutput;
import gg.jte.runtime.TemplateUtils;
import org.assertj.core.api.AbstractThrowableAssert;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import test.Model;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.catchThrowable;

/**
 * To run these tests, you first need to run the jte-maven-plugin (jte:precompile)
 */
public class TemplateEngineTest {

    private static TemplateEngine templateEngine;
    StringOutput output = new StringOutput();
    Model model = new Model();

    @BeforeAll
    static void beforeAll() {
        templateEngine = TemplateEngine.createPrecompiled(null, ContentType.Html, null, "gg.jte.additionalgeneratetask");
    }

    @BeforeEach
    void setUp() {
        model.hello = "Hello";
        model.x = 42;
    }

    @Test
    void helloWorld() {
        whenTemplateIsRendered("helloWorld.jte");
        thenOutputIs("Hello World");
    }

    @Test
    void templateNotFound() {
        thenRenderingFailsWithException("unknown.jte").hasMessage("Failed to load unknown.jte");
    }

    @Test
    void exceptionLineNumber1() {
        thenRenderingFailsWithException("exceptionLineNumber1.jte")
                .hasCauseInstanceOf(NullPointerException.class)
                .hasMessage("Failed to render exceptionLineNumber1.jte, error at exceptionLineNumber1.jte:5");
    }

    @Test
    void unusedTag() {
        templateEngine.render("tag/unused.jte", TemplateUtils.toMap("param1", "One", "param2", "Two"), output);
        thenOutputIs("One is One, two is Two.");
    }

    @Test
    void params() {
        Throwable throwable = catchThrowable(() -> templateEngine.getParamInfo("tag/unused.jte"));
        assertThat(throwable).isInstanceOf(TemplateException.class).hasMessage("No parameter information is available for tag/unused.jte, compile templates with -parameters flag, to use this method.");
    }

    @Test
    void paramWithWrongType() {
        Throwable throwable = catchThrowable(() -> templateEngine.render("helloWorld.jte", TemplateUtils.toMap("model", "string"), output));
        assertThat(throwable).isInstanceOf(TemplateException.class).hasMessageContaining("Failed to render helloWorld.jte, error at helloWorld.jte:1");
    }

    private void whenTemplateIsRendered(String templateName) {
        templateEngine.render(templateName, model, output);
    }

    private AbstractThrowableAssert<?, ? extends Throwable> thenRenderingFailsWithException(String templateName) {
        Throwable throwable = catchThrowable(() -> whenTemplateIsRendered(templateName));
        return assertThat(throwable).isInstanceOf(TemplateException.class);
    }

    private void thenOutputIs(String expected) {
        assertThat(output.toString()).isEqualTo(expected);
    }
}

File: test/jte-runtime-cp-test-gradle-kotlin/src/test/kotlin/test/TemplateEngineTest.kt
package test

import gg.jte.ContentType
import gg.jte.TemplateEngine
import gg.jte.TemplateException
import gg.jte.output.StringOutput
import gg.jte.runtime.TemplateUtils
import org.junit.jupiter.api.*
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import java.lang.NullPointerException

/**
 * To run these tests, you first need to run the jte-gradle-plugin (jte:precompile)
 */
class TemplateEngineTest {

    companion object {
        val templateEngine = TemplateEngine.createPrecompiled(ContentType.Html)
    }

    val output = StringOutput()
    val model = Model()

    @BeforeEach
    internal fun setUp() {
        model.hello = "Hello"
    }

    @Test
    internal fun helloWorld() {
        whenTemplateIsRendered("helloWorld.kte")
        thenOutputIs("Hello World")
    }

    @Test
    internal fun templateNotFound() {
        val exception = thenRenderingFailsWithException("unknown.kte")
        assertEquals("Failed to load unknown.kte", exception.message)
    }

    @Test
    internal fun exceptionLineNumber1() {
        val exception = thenRenderingFailsWithException("exceptionLineNumber1.kte")
        assertTrue(exception.cause is NullPointerException)
        assertEquals("Failed to render exceptionLineNumber1.kte, error at exceptionLineNumber1.kte:5", exception.message)
    }

    @Test
    internal fun unusedTag() {
        templateEngine.renderTag("tag/unused.kte", TemplateUtils.toMap("param1", "One", "param2", "Two"), output)
        thenOutputIs("One is One, two is Two.")
    }

    @Test
    internal fun params() {
        val exception = assertThrows<TemplateException> { templateEngine.getParamInfo("tag/unused.kte") }
        assertEquals("No parameter information is available for tag/unused.kte, compile templates with -parameters flag, to use this method.", exception.message)
    }

    @Test
    internal fun paramWithWrongType() {
        val exception = assertThrows<TemplateException> { templateEngine.render("helloWorld.kte", TemplateUtils.toMap("model", "string"), output) }
        assertTrue(exception.message!!.contains("Failed to render helloWorld.kte, error at helloWorld.kte:0"))
    }

    private fun whenTemplateIsRendered(templateName: String) {
        templateEngine.render(templateName, model, output)
    }

    private fun thenRenderingFailsWithException(templateName: String): TemplateException {
        return assertThrows { whenTemplateIsRendered(templateName) }
    }

    private fun thenOutputIs(expected: String) {
        assertEquals(expected, output.toString())
    }
}